### PR TITLE
Add tests for new JuMP nonlinear expression syntax

### PR DESCRIPTION
**Pull request by odow**
x-ref https://github.com/jump-dev/JuMP.jl/pull/3106

This found a couple of bugs in my Ipopt.jl implementation, but now things are all passing :smile:

### CODE REVIEW

**Comment by ccoffrin** [comment]
Test FTW!

File: src/nlp-mi/005_011.jl [hunk 1]
```diff
@@ -15,9 +15,9 @@ function nlp_mi_005_011(
     @variable(model, y >= 0)
 
     @objective(model, Min, x + y)
-    @NLconstraint(model, y >= (x + 0.1) \ 1 - 0.5)
-    @NLconstraint(model, x >= y^(-2) - 0.5)
-    @NLconstraint(model, (x + y + 0.1) \ 4 >= 1)
+    @constraint(model, y >= (x + 0.1) \ 1 - 0.5)
+    @constraint(model, x >= y^(-2) - 0.5)
+    @constraint(model, (x + y + 0.1) \ 4 >= 1)
```
Comment (odow) [inline src/nlp-mi/005_011.jl:20]: We're not testing this because of `\`. How often is `\` actually used in a scalar NLP?

Comment (odow) [reply src/nlp-mi/005_011.jl:20]: This was never supported by JuMP, as far as I know 😆

Comment (ccoffrin) [reply src/nlp-mi/005_011.jl:20]: Ok maybe it was more about enumerating every basic function that is supported by Julia. I am find to remove it if you think it is not reasonable.

Comment (odow) [reply src/nlp-mi/005_011.jl:20]: It's skipped in the tests so not a big deal. We can leave as-is. I wondered if there was some particular application you had in mind.

File: test/runtests.jl [hunk 2]
```diff
@@ -1,3 +1,7 @@
+import Pkg
+Pkg.pkg"add JuMP#od/nlp-expr"
+Pkg.pkg"add MathOptInterface#master"
```
Comment (odow) [inline test/runtests.jl:3]:
```suggestion
Pkg.pkg"add MathOptInterface#od/fix-nl-saf"
```

File: test/runtests.jl [hunk 3]
```diff
@@ -1,3 +1,7 @@
+import Pkg
+Pkg.pkg"add JuMP#od/nlp-expr"
+Pkg.pkg"add MathOptInterface#od/fix-nl-saf"
```
Comment (odow) [inline test/runtests.jl:3]:
```suggestion
Pkg.pkg"add MathOptInterface#master"
```

File: test/runtests.jl [hunk 4]
```diff
@@ -1,3 +1,6 @@
+import Pkg
+Pkg.pkg"add JuMP#master"
```
Comment (odow) [inline test/runtests.jl:2]: TODO: remove once tagged

File: Project.toml [hunk 5]
```diff
@@ -1,21 +1,22 @@
...
 JuMP = "4076af6c-e467-56ae-b986-b466b2749572"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [compat]
-Ipopt = "0.6, 0.7, 0.8, 1"
-JuMP = "0.21, 0.22, 0.23, 1"
-Juniper = "0.7, 0.8"
-julia = "1"
+Ipopt = "1"
+JuMP = "1"
```
Comment (odow) [inline Project.toml:12]: TODO: update once tagged

File: Project.toml [hunk 6]
```diff
@@ -1,21 +1,22 @@
...
+Juniper = "0.9"
+julia = "1.6"
 
 [extras]
 Ipopt = "b6b21f68-93f8-5de0-b562-5493be1d77c9"
 Juniper = "2ddba703-00a4-53a7-87a5-e8b9971dde84"
+Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
```
Comment (odow) [inline Project.toml:19]: TODO: remove once tagged

**Comment by odow** [comment]
Carleton suggests adding these tests separately so that we can test both nonlinear interfaces to JuMP

**Comment by odow** [comment]
This is now opt-in for solvers. The old functionality is unchanged.

**Pull request merged**